### PR TITLE
Grab the firmware implementation of logging over SBP.

### PR DESCRIPTION
**Pull request by kmdade**
This replaces the previous behavior of using `popen` with the `sbp_log` application for every log.

### CODE REVIEW

**Comment by silverjam** [comment]
FYI, there's some build errors too:

```
logging.c: In function 'logging_init':
logging.c:34:15: error: 's' undeclared (first use in this function)
   if (NULL == s->sbp_tx) {
               ^
logging.c:34:15: note: each undeclared identifier is reported only once for each function it appears in
logging.c: In function 'sbp_log':
logging.c:82:12: error: 'level' undeclared (first use in this function)
   sbp_vlog(level, msg_text, ap);
            ^~~~~
logging.c: At top level:
logging.c:86:6: error: conflicting types for 'sbp_vlog'
 void sbp_vlog(int priority, const char *msg, ...)
      ^~~~~~~~
In file included from logging.c:13:0:
../include/libpiksi/logging.h:97:6: note: previous declaration of 'sbp_vlog' was here
 void sbp_vlog(int priority, const char *msg_text, va_list ap);
      ^~~~~~~~
make[5]: *** [logging.o] Error 1
Makefile:18: recipe for target 'logging.o' failed
```

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v2.0.0-develop-2018102218-11-g2c9518a4
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=738625)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018102218-11-g2c9518a4&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018102218-11-g2c9518a4&groupby_key=firmware&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v2.0.0-develop-2018102218-11-g2c9518a4
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.0.0-develop-2018102218-11-g2c9518a4)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.0.0-develop-2018102218-11-g2c9518a4/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.0.0-develop-2018102218-11-g2c9518a4

File: package/libpiksi/libpiksi/src/logging.c [hunk 1]
```diff
@@ -73,39 +83,29 @@ void sbp_log(int priority, const char *msg_text, ...)
...
-  vsnprintf(formatted_msg, sizeof(formatted_msg), msg_text, ap);
+  log = (msg_log_t *)buf;
+  log->level = (uint8_t)priority;
 
-  char msg_buf[2048];
-  snprintf(msg_buf, sizeof(msg_buf), "%s: %s", log_ident, formatted_msg);
+  int n = vsnprintf(log->text, SBP_FRAMING_MAX_PAYLOAD_SIZE - sizeof(msg_log_t), msg, ap);
 
-  fputs(msg_buf, output);
+  if (n < 0) return;
```
Comment (yizow) [inline package/libpiksi/libpiksi/src/logging.c]: Do we want an explicit error logged in this case?

**Comment by silverjam** [comment]
This doesn't boot (or at least appears to not boot because the tcp/uart adapters never come up)... and you'll never guess why!
.
.
.
.
.
It's because nanomsg doesn't work across fork() calls (and fails by apparently causing the forked child to block when attempting to sent to a socket that it's parent has already connected to).  Since endpoint_adapter uses forking to setup it's send/recv loops it causes the endpoint_adapter to stall.

File: package/libpiksi/libpiksi/src/logging.c (hunk 1 above)
Comment (silverjam) [reply package/libpiksi/libpiksi/src/logging.c]: I think we'll be OK without it, at the point where we have code that needs to react with sbp_log fails we can add a return code.

**Comment by silverjam** [comment]
@kmdade May want to revisit this line in firmware:

```
int n = vsnprintf(log->text, SBP_FRAMING_MAX_PAYLOAD_SIZE - sizeof(msg_log_t), msg, ap);
```

The `vsnprintf` function returns the number bytes "that would have been printed if the buffer were of unlimited" size-- so the line that uses "n" (without modification):

```
if (0 != sbp_tx_send(sbp_tx, SBP_MSG_LOG, n + sizeof(msg_log_t), (uint8_t *)buf)) {
```

Is technically telling the `sbp_tx_send` function to read past the end of the buffer.

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v2.0.0-develop-2018102218-13-g00b7500e
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=739263)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018102218-13-g00b7500e&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018102218-13-g00b7500e&groupby_key=firmware&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v2.0.0-develop-2018102218-13-g00b7500e
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.0.0-develop-2018102218-13-g00b7500e)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.0.0-develop-2018102218-13-g00b7500e/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.0.0-develop-2018102218-13-g00b7500e

**Comment by benjaminaltieri** [comment]
Can we get a JIRA ticket added to cover the reasoning behind and effect of this PR? It's a relatively small and straightforward change but would be good to have traceability.

**Comment by silverjam** [comment]
https://swift-nav.atlassian.net/browse/DEVC-1250

**Pull request merged**